Add memory functions like malloc, free, memset, zero_memory. And basic user input support.

izanbf1803 opened the pull request.

izanbf1803 committed:
Implemented memory functions kmalloc, kzero_memory, kfree
Also added a struct (memory_block) to manage memory segments

izanbf1803 committed:
Added user input support
Implemented RETN key (delete last char pressed by user)
Implemented multiple key modificators like BLOQ_MAYUS, SHIFT, and ALT_GR

izanbf1803 committed:
Added RETN lock after printk()

izanbf1803 committed:
Memory allocation and deallocation errors solved.

izanbf1803 committed:
Merge branch 'master' into master

danirod commented:
This will require manual intervention because of the conflicts it is given. A lot of things have changed since the first commit contained in this PR. To avoid this in the future, whenever you are working in forks and you let your branches stale you should add the original repository as an additional remote and time by time make some fetches from upstream to receive the new changes made by other authors. [Check this out][1] for further information.

I'm interested in the memory functions. The memory management will probably have to change in the future anyway but this is better than nothing. I don't know what will happen to the VGA console because the code is moving towards VBE to support graphical modes.

[1]: https://robots.thoughtbot.com/keeping-a-github-fork-updated

izanbf1803 committed:
Removed all functions related to VGA screen and keyboard, added memory functions
Also added macros to support printk() with format and output to the serial port if DEBUG is set to True

izanbf1803 committed:
Merge branch 'master' of https://github.com/izanbf1803/nativeos

izanbf1803 committed:
Added DEBUG option to VGA functions (redirect output to serial port)

danirod commented on kernel/driver/keyboard_map.c:
```diff
@@ -0,0 +1,166 @@
+	keyboard_dict[36] = 'j';
+	keyboard_dict[37] = 'k';
+	keyboard_dict[38] = 'l';
+	keyboard_dict[39] = 'ñ';
```
Compiler warns here because extended ASCII characters (>127 such as ñ or ç) are being stored in a signed char. Might not cause regressions as long as values are never treated as numbers, only as characters, but it may be better to just make keyboard_dict arrays unsigned char.

izanbf1803 replied:
Solved.

danirod commented on kernel/memory.c:
```diff
@@ -69,49 +99,48 @@ static unsigned char* next_address = &kernel_after;
-    next_address += size;
-    return address_to_return;
-}
+void* kmalloc(unsigned int size)
```
The thing with kmalloc_real is that sometimes you need to align pages to a 4 KB memory page (this is, you really want your allocated memory address to end by 0x000), and that is the purpose of the `align` flag. I don't know if this new implementation will align buffers into page boundaries. I don't remember a lot about this at this moment either unless I review my technical documents, which I don't know where they are.

danirod commented on kernel/main.c:
```diff
@@ -1,6 +1,10 @@
 /*
  * This file is part of NativeOS: next-gen x86 operating system
- * Copyright (C) 2015-2017 Dani Rodríguez
+<<<<<<< HEAD
```
I see you've had a bad experience with git merge or git pull here... 😛

izanbf1803 replied:
Oh haha, solved.

danirod commented on kernel/driver/keyboard.c:
```diff
@@ -24,17 +24,73 @@
 #include <arch/x86/io.h>
 #include <driver/keyboard.h>
 
+char shift_pressed = 0;
```
If these variables aren't used outside keyboard.c they could be made static char so that the identifiers aren't exported. There could be conflicts if two .o files export the same variable identifiers during the linking stage, even if they are unrelated, so keeping things `static` unless they are really required to be used from a different source code file mitigates this risk.

danirod reviewed (changes requested):
There are a few issues regarding variable types that I would like to have addressed before adding this in keyboard.c and keyboard_map.c.

I don't mind having breaking changes in memory.c because old memory.c implementation was dumb (it was a simple bitmap allocator anyway). However, that code will probably evolve the day I come back to this project and start working on an MMU because page alignment during malloc will probably be a requirement at that point.

I don't mind keeping the code for keyboard and VGA at the moment. Kernel was never the place for a keyboard console in the first place, but it is always fun to see the screen do things when you press keys. Usually the keyboard driver only collects those scancodes and exposes them so that userland applications can ask "hey kernel what key did the user just press?" and handle it.

Similar thing goes for VGA. I'm considering keeping vga.c even if it's not used when the graphics mode is added in case I add a framebuffer –a virtual console on top of a graphics mode.

danirod commented on Makefile:
```diff
@@ -40,7 +40,7 @@ INCLUDE_PATH = include
 
 # Tool flags
-CFLAGS = -nostdlib --freestanding -fno-builtin -g -I$(INCLUDE_PATH)/
+CFLAGS = -nostdlib --freestanding -fno-builtin -g -I $(INCLUDE_PATH)/
```
This isn't actually a bug. GCC won't care whether you have a space after -I or not.

In fact, until today I thought you were supposed not to, because I have always seen this flag without any spaces; but, because this compiles, I guess it's actually legal to have a space there.

izanbf1803 committed:
Solved some little coding errors.

danirod commented:
Disclaimer to future visitors: this PR is approved because it was submitted before the new [contributing rules](https://github.com/danirod/nativeos/blob/master/CONTRIBUTING.md#pull-requests) (which rejects PR by default) was approved and I considered unpolite to force this retroactively to already existing PRs.

_Disclaimer para futuros visitantes: este PR ha sido aprobado porque fue enviado antes de establecer las nuevas [reglas de contribución](https://github.com/danirod/nativeos/blob/master/CONTRIBUTING.md#pull-requests) (que básicamente rechazan los PRs), las cuales consideré poco educado aplicar de forma retroactiva a PRs existentes.
